tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -1535,6 +1535,127 @@ class StaticAlgorithmSelector : public nvinfer1::IAlgorithmSelector {
+          intersection.emplace_back(allowed);
+        }
+      }
+      if (intersection.size() > 0) {
+        VLOG(1) << "Intersection of allowed and disallowed tactics found be "
+                   "non-empty.";
+        VLOG(1) << "List of allowed tactcis:";
+        for (int64_t a : allowed_tactics) {
+          VLOG(1) << a << ", ";
+        }
+        VLOG(1) << "List of disallowed tactcis:";
+        for (int64_t d : disallowed_tactics) {
+          VLOG(1) << d << ", ";
+        }
+        VLOG(1) << "List of tactics common to both allowed and disallowed "
+                   "tactics:";
+        for (int64_t i : intersection) {
+          VLOG(1) << i << ", ";
+        }
+        VLOG(1) << "Aborting.";
+        assert(0);
+      }
+    }
+  }
+  // Returns value in [0, nbChoices] for a valid algorithm.
+  int32_t selectAlgorithms(const nvinfer1::IAlgorithmContext& algoContext,
+                           const nvinfer1::IAlgorithm* const* algoChoices,
+                           int32_t nbChoices, int32_t* selection) override {
+    bool reject{false};
+    // TensorRT always provides more than zero number of algorithms
+    // in selectAlgorithms.
+    assert(nbChoices > 0);
+
+    if (allowed_tactics.empty()) {
+      // Disable all disallowed tactics
+      int32_t nbSelections = 0;
+      for (auto i = 0; i < nbChoices; i++) {
+        if (algoChoices[i] == nullptr) {
+          continue;
+        }
+
+        int64_t tacticId = algoChoices[i]->getAlgorithmVariant().getTactic();
+        int64_t implementation =
+            algoChoices[i]->getAlgorithmVariant().getImplementation();
+        nvinfer1::TensorFormat format =
+            algoChoices[i]->getAlgorithmIOInfo(0).getTensorFormat();
+        assert(tacticId != static_cast<int64_t>(Tactic::kINVALID_TACTIC));
+
+        if (disallowed_tactics.find(tacticId) != disallowed_tactics.end()) {
+          reject = true;
+        }
+
+#if !IS_TRT_VERSION_GE(8, 0, 0, 0)
+        if (implementation == static_cast<int64_t>(LayerImpl::kSHUFFLE) &&
+            format == nvinfer1::TensorFormat::kCHW32) {
+          reject = true;
+        }
+#endif
+
+        if (reject) {
+          VLOG(1) << "Rejecting a disallowed tactic: " << tacticId
+                  << " for node " << algoContext.getName()
+                  << " with implementation " << implementation
+                  << " and input format " << static_cast<int32_t>(format);
+        } else {
+          selection[nbSelections++] = i;
+        }
+      }
+      return nbSelections;
+    } else {
+      assert(allowed_tactics.size() == 1);
+      int64_t algorithm_id = *allowed_tactics.begin(0);
+      VLOG(1) << "Forcing TRT algorithm selection to: ID=" << algorithm_id;
+      // making sure that the requested TRT algorithm ID doesn't go above the
+      // max value accepted.
+      selection[0] = std::min(algorithm_id, static_cast<int64_t>(nbChoices));
+      return 1;
+    }
+  }
```

Comment:
When we set TF_TRT_FIXED_ALGORITHM_ID we still want to exclude SHUFFLE for the format with a bug. This is not implemented in code right now. Here is my suggestion to re-flow the whole routine to fix this issue and also simplify the logic:

nbChoices > 0
if both allowed_tactics and disallowed_tactics are empty return nbChoices

for each i in 0..nbChoices
    if the implementation is for the SHUFFLE format with bug, reject it, continue to the next iteration

    if (allowed_tactics is not empty) {
        if the tactics id is in allowed_tactics, select it
        otherwise, reject it
        continue to the next iteration
   }

    if the tactics id is NOT in disallowed_tactics, select it
    otherwise reject it

Reply:
@jhalakpatel please take a look ;)